### PR TITLE
Userland: Add ability to screenshot rectangular region in `shot`

### PR DESCRIPTION
**Pull request by abyesilyurt**
This PR adds a new option to shot to take screenshots of rectangular regions.
User is shown a semi-transparent overlay like below, where s/he can select a rectangle with mouse clicks.

- I had to change the way fullscreen transparent windows are drawn. 
Before that, all the windows in the background were ignored in the existance of a fullscreen window.
- ~`Bitmap::cropped` function used to return a pointer to local variable,
and the returned bitmap was invalid.
I fixed it by using `static`, but there should be a better way.~
- The taskbar is not selectable (yet), related issue #8510.

Screenshot_20210707_012653

### CODE REVIEW

File: Userland/Libraries/LibGfx/Bitmap.cpp [hunk 1]
```diff
@@ -476,7 +476,7 @@ RefPtr<Gfx::Bitmap> Bitmap::scaled(float sx, float sy) const
 
 RefPtr<Gfx::Bitmap> Bitmap::cropped(Gfx::IntRect crop) const
 {
-    auto new_bitmap = Gfx::Bitmap::create(format(), { crop.width(), crop.height() }, 1);
+    static auto new_bitmap = Gfx::Bitmap::create(format(), { crop.width(), crop.height() }, 1);
```
Comment (alimpfard) [inline Userland/Libraries/LibGfx/Bitmap.cpp:479]: This is not necessary, `new_bitmap` is a RefPtr which will be copied on return.

File: Userland/Services/WindowServer/Compositor.cpp [hunk 2]
```diff
@@ -1096,7 +1097,8 @@ void Compositor::recompute_occlusions()
         });
 
         m_opaque_wallpaper_rects.clear();
-    } else {
+    }
+    {
```
Comment (alimpfard) [inline Userland/Services/WindowServer/Compositor.cpp:1101]: I'm not quite sure why this (massive) else branch is now always being run?

File: Userland/Libraries/LibGfx/Bitmap.cpp (hunk 1 above)
Comment (gunnarbeutner) [inline Userland/Libraries/LibGfx/Bitmap.cpp:479]: That doesn't look correct. For one this would return the same bitmap even if - at a later point - a different IntRect was specified.

Also, this seems unnecessary because `new_bitmap` is a heap-allocated object that's wrapped in a smart pointer (RefPtr).

Instead you should change `auto* bitmap` to `auto bitmap` in `shot.cpp` because that would otherwise destroy the RefPtr.

File: Userland/Utilities/shot.cpp [hunk 3]
```diff
@@ -5,28 +5,100 @@
...
     bool output_to_clipboard = false;
     int delay = 0;
+    bool select_region = false;
     int screen = -1;
 
     args_parser.add_positional_argument(output_path, "Output filename", "output", Core::ArgsParser::Required::No);
     args_parser.add_option(output_to_clipboard, "Output to clipboard", "clipboard", 'c');
     args_parser.add_option(delay, "Seconds to wait before taking a screenshot", "delay", 'd', "seconds");
     args_parser.add_option(screen, "The index of the screen (default: -1 for all screens)", "screen", 's', "index");
+    args_parser.add_option(select_region, "Select a region to capture", "region", 'r');
```
Comment (alimpfard) [inline Userland/Utilities/shot.cpp:101]: It'd also be neat to make this scriptable, so you can say `shot --rect 10,10,40,40` or something to that effect... though that can be done in a later PR/change too.

File: Userland/Utilities/shot.cpp [hunk 4]
```diff
@@ -5,28 +5,100 @@
...
+            abs(p1.value().x() - p2.value().x()),
+            abs(p1.value().y() - p2.value().y()));
+
+        return Gfx::IntRect(point, size);
+    }
+
+    Optional<Gfx::IntPoint> m_last_point {};
+    Optional<Gfx::IntPoint> m_achor_point {};
+    Optional<Gfx::IntRect> m_region {};
+    GUI::Window* m_window;
```
Comment (alimpfard) [inline Userland/Utilities/shot.cpp:84]: Also, please initialise primitive values.
(And Optional<T> has a default ctor, no need to initialise it manually)

**Review by gunnarbeutner — CHANGES_REQUESTED** [review]

File: Userland/Utilities/shot.cpp [hunk 5]
```diff
@@ -5,28 +5,100 @@
...
 #include <LibGUI/Application.h>
 #include <LibGUI/Clipboard.h>
+#include <LibGUI/Painter.h>
+#include <LibGUI/Widget.h>
+#include <LibGUI/Window.h>
 #include <LibGUI/WindowServerConnection.h>
 #include <LibGfx/PNGWriter.h>
 #include <unistd.h>
 
+class SelectableLayover final : public GUI::Widget {
```
Comment (alimpfard) [inline Userland/Utilities/shot.cpp:21]: It feels rather awkward to do click-click or click-click+drag to select a region, why not just have a normal region select mechanism similar to pixelpaint? i.e. click+drag to select, and release to take a screenshot?

**Comment by gunnarbeutner** [comment]
Another thing I just noticed: Can you make it so that the selected area isn't grayed out like the rest? :)

File: Userland/Utilities/shot.cpp [hunk 6]
```diff
@@ -5,28 +5,100 @@
...
+    {
+        if (event.key() != Key_Return)
+            return;
+
+        m_window->close();
+    }
+
+    static Optional<Gfx::IntRect> build_rect(Optional<Gfx::IntPoint> const& p1, Optional<Gfx::IntPoint> const& p2)
+    {
+        if (!p1.has_value() || !p1.has_value())
```
Comment (ChrisK92) [inline Userland/Utilities/shot.cpp:68]: You probably want:
` if (!p1.has_value() || !p2.has_value())`

File: Userland/Utilities/shot.cpp [hunk 7]
```diff
@@ -5,28 +5,100 @@
...
+        m_last_point = event.position();
+        if (m_achor_point.has_value())
+            update();
+    };
+
+    virtual void paint_event(GUI::PaintEvent&) override
+    {
+        m_region = build_rect(m_achor_point, m_last_point);
+        GUI::Painter painter(*this);
+        painter.draw_rect(m_region.value_or(Gfx::IntRect()), Gfx::Color::Black);
```
Comment (MaxWipfli) [inline Userland/Utilities/shot.cpp:55]: The `build_rect` function basically already exists in `Rect<T>`, so you can do this instead of writing your own function. Also, don't instantiate a `GUI::Painter` to draw an zero-size rect.
```suggestion
        if (!m_achor_point.has_value() || !m_last_point.has_value()) {
            m_region = {};
            return;
        }
        GUI::Painter painter(*this);
        m_region = IntRect::from_two_points(*m_achor_point, *m_last_point);
```

File: Userland/Utilities/shot.cpp [hunk 8]
```diff
@@ -5,28 +5,100 @@
...
+            min(p1.value().y(), p2.value().y()));
+        Gfx::Size size(
+            abs(p1.value().x() - p2.value().x()),
+            abs(p1.value().y() - p2.value().y()));
+
+        return Gfx::IntRect(point, size);
+    }
+
+    Optional<Gfx::IntPoint> m_last_point {};
+    Optional<Gfx::IntPoint> m_achor_point {};
```
Comment (MaxWipfli) [inline Userland/Utilities/shot.cpp:82]: I guess this should be `m_anchor_point`.

File: Userland/Utilities/shot.cpp [hunk 9]
```diff
@@ -35,6 +107,18 @@ int main(int argc, char** argv)
...
+    auto window = GUI::Window::construct();
+    Optional<Gfx::IntRect> crop_region;
+    if (select_region) {
+        window->set_opacity(0.4);
+        auto& container = window->set_main_widget<SelectableLayover>(window);
+        container.set_fill_with_background_color(true);
+        window->set_fullscreen(true);
+        window->show();
+        app->exec();
+        crop_region = container.region();
```
Comment (MaxWipfli) [inline Userland/Utilities/shot.cpp:119]: I guess if there is no `crop_region`, the user has aborted and the program should thus exit? Currently, it will then just make a screenshot of the full screen.

File: Userland/Utilities/shot.cpp [hunk 10]
```diff
@@ -5,28 +5,100 @@
...
+    virtual void paint_event(GUI::PaintEvent&) override
+    {
+        m_region = build_rect(m_achor_point, m_last_point);
+        GUI::Painter painter(*this);
+        painter.draw_rect(m_region.value_or(Gfx::IntRect()), Gfx::Color::Black);
+    }
+
+    virtual void keydown_event(GUI::KeyEvent& event) override
+    {
+        if (event.key() != Key_Return)
```
Comment (MaxWipfli) [inline Userland/Utilities/shot.cpp:60]: I didn't test this, but I would expect pressing Escape to abort the screenshot process. I think this is not the case at the moment, right?

**Review by MaxWipfli — CHANGES_REQUESTED** [review]
A couple of comments. Also, please make the selected area non-grayed out, and make it so that it's Click+Drag to select, like what PixelPaint does (as Ali and Gunnar already mentioned).

File: Userland/Libraries/LibGfx/Bitmap.cpp (hunk 1 above)
Comment (abyesilyurt) [reply Userland/Libraries/LibGfx/Bitmap.cpp]: Reverted the change.

File: Userland/Services/WindowServer/Compositor.cpp (hunk 2 above)
Comment (abyesilyurt) [reply Userland/Services/WindowServer/Compositor.cpp]: I changed this line again, can you have a look? The idea is basically to keep drawing the background, when the fullscreen window is transparent.

File: Userland/Utilities/shot.cpp (hunk 4 above)
Comment (abyesilyurt) [reply Userland/Utilities/shot.cpp]: Does it look better now?

File: Userland/Services/WindowServer/Compositor.cpp (hunk 2 above)
Comment (alimpfard) [reply Userland/Services/WindowServer/Compositor.cpp]: looks like it does what you're saying it does, so that's alright by me :P

File: Userland/Utilities/shot.cpp (hunk 4 above)
Comment (alimpfard) [reply Userland/Utilities/shot.cpp]: yup yup

File: Userland/Libraries/LibGfx/Bitmap.cpp (hunk 1 above)
Comment (abyesilyurt) [reply Userland/Libraries/LibGfx/Bitmap.cpp]: Thanks for the correction! 

Can you help me understand why `auto* bitmap` would destroy `RefPtr`?

**Comment by abyesilyurt** [comment]
> Another thing I just noticed: Can you make it so that the selected area isn't grayed out like the rest? :)

Right now, all this stuff seems to be happening in `Compositor`, which is hard to follow for me.
Is it possible to implement that without any change in the  `Compositor` logic?

**Comment by MaxWipfli** [comment]
> > Another thing I just noticed: Can you make it so that the selected area isn't grayed out like the rest? :)
> 
> Right now, all this stuff seems to be happening in `Compositor`, which is hard to follow for me as a new contributor.
> Is it possible to implement that without any change in the `Compositor` logic?

I think you need to use `fill_rect` instead of `draw_rect`, and then use `Color::Transparent` as the color.

File: Userland/Libraries/LibGfx/Bitmap.cpp (hunk 1 above)
Comment (gunnarbeutner) [reply Userland/Libraries/LibGfx/Bitmap.cpp]: `Gfx::Bitmap::create` returns a `RefPtr<Gfx::Bitmap>`. When you use `auto*` that causes C++ to try to coerce whatever object you specified to the right of `=` into a pointer type.

The `RefPtr<T>` class has this operator which is then implicitly used to coerce the `RefPtr<Gfx::Bitmap>` into a `Gfx::Bitmap*`: `ALWAYS_INLINE operator T*() { return as_ptr(); }`

However, at that point the `RefPtr<Gfx::Bitmap>` that was returned by the `create()` function is no longer necessary and is implicitly destroyed (because you're not holding on to it anywhere) - which in turn also destroys the underlying `Gfx::Bitmap` object - leaving you with a dangling pointer.

File: Userland/Utilities/shot.cpp (hunk 9 above)
Comment (abyesilyurt) [reply Userland/Utilities/shot.cpp:115]: If there is no `crop_region`, the program will just exit now.

File: Userland/Utilities/shot.cpp (hunk 10 above)
Comment (abyesilyurt) [reply Userland/Utilities/shot.cpp]: Yes, you are right. I added Escape option to quit the program without any screenshots.

File: Userland/Utilities/shot.cpp (hunk 7 above)
Comment (abyesilyurt) [reply Userland/Utilities/shot.cpp]: Thanks! Applied the changes.

File: Userland/Utilities/shot.cpp (hunk 6 above)
Comment (abyesilyurt) [reply Userland/Utilities/shot.cpp]: Good catch :) but not necessary anymore as @MaxWipfli pointed to `Gfx::IntRect::from_two_points`.

**Comment by abyesilyurt** [comment]
> > > Another thing I just noticed: Can you make it so that the selected area isn't grayed out like the rest? :)
> > 
> > 
> > Right now, all this stuff seems to be happening in `Compositor`, which is hard to follow for me as a new contributor.
> > Is it possible to implement that without any change in the `Compositor` logic?
> 
> I think you need to use `fill_rect` instead of `draw_rect`, and then use `Color::Transparent` as the color. Although maybe it's required to make the full window transparent, and then fill everything but the rect with some semi-transparent color.

I gave it a try, but didn't work. I think Color::Transparent only works across the widgets within the same window, i.e. you see the bottom of the window.

I also tried to make the window transparent, in that case `draw_rect` did not have any effect. Is it possible to draw on transparent windows?

File: Userland/Utilities/shot.cpp (hunk 5 above)
Comment (abyesilyurt) [reply Userland/Utilities/shot.cpp:22]: Switched to click+drag to select, the code is also much cleaner now :)

File: Userland/Libraries/LibGfx/Bitmap.cpp (hunk 1 above)
Comment (abyesilyurt) [reply Userland/Libraries/LibGfx/Bitmap.cpp]: Thanks for the detailed info! Everything is much clear now.

File: Userland/Utilities/shot.cpp (hunk 3 above)
Comment (abyesilyurt) [reply Userland/Utilities/shot.cpp:90]: Good idea! I will look into that in a later PR. Which one makes more sense to define the rect x,y,w,h or x1,y1,x2,y2?

**Comment by abyesilyurt** [comment]
Thanks for all the input everyone! 

I addressed all the comments except non-graying out the selected region. I believe it is more complicated than it sounds.

File: Userland/Utilities/shot.cpp (hunk 3 above)
Comment (alimpfard) [reply Userland/Utilities/shot.cpp:90]: I personally like x,y,w,h but we can bikeshed that when needed :P

**Comment by alimpfard** [comment]
To make it not draw the selected rect, you could perhaps try not filling the window with the bg colour, and instead manually filling the regions that are meant to be grayed out, in pseudocode:
```
paint_event()
    fill_with_transparency()
    for each rect in window_rect.shatter(selected_rect):
        fill(rect, White)
    draw_rectangle(selected_rect, Black)
```

File: Userland/Utilities/shot.cpp [hunk 11]
```diff
@@ -1,32 +1,93 @@
...
+#include <LibGUI/Window.h>
 #include <LibGUI/WindowServerConnection.h>
 #include <LibGfx/PNGWriter.h>
 #include <unistd.h>
 
+class SelectableLayover final : public GUI::Widget {
+    C_OBJECT(SelectableLayover)
+public:
+    SelectableLayover(GUI::Window* window)
+        : m_window(window) {};
```
Comment (alimpfard) [inline Userland/Utilities/shot.cpp:26]: Please move the empty block to a new line, and remove the extraneous `;`
```suggestion
    SelectableLayover(GUI::Window* window)
        : m_window(window)
    {
    }
```

**Comment by abyesilyurt** [comment]
> To make it not fill the selected rect, you could perhaps try not filling the window with the bg colour, and instead manually filling the regions that are meant to be grayed out, in pseudocode:
> 
> ```
> paint_event()
>     fill_with_transparency()
>     for each rect in window_rect.shatter(selected_rect):
>         fill(rect, White)
>     draw_rectangle(selected_rect, Black)
> ```

Works!
image

File: Userland/Utilities/shot.cpp [hunk 12]
```diff
@@ -1,32 +1,103 @@
...
+    {
+        if (event.button() == GUI::MouseButton::Left)
+            m_window->close();
+    };
+
+    virtual void paint_event(GUI::PaintEvent&) override
+    {
+        if (m_region.is_empty()) {
+            GUI::Painter painter(*this);
+            painter.fill_rect(m_window->rect(), m_background_color);
```
Comment (abyesilyurt) [inline Userland/Utilities/shot.cpp:62]: Is it correct to repaint the background everytime this function is called?

File: Userland/Utilities/shot.cpp (hunk 11 above)
Comment (abyesilyurt) [reply Userland/Utilities/shot.cpp]: Done

File: Userland/Utilities/shot.cpp (hunk 12 above)
Comment (alimpfard) [reply Userland/Utilities/shot.cpp:62]: Eh, why not :shrug:

**Comment by alimpfard** [comment]
that image looks a little inverted though :thinking: 
shouldn't the non-selected area be grayed out?

File: Userland/Utilities/shot.cpp (hunk 12 above)
Comment (abyesilyurt) [reply Userland/Utilities/shot.cpp:62]: I don't know much about the internals, I thought it might be costly to call everytime.

**Comment by abyesilyurt** [comment]
> that image looks a little inverted though 
> shouldn't the non-selected area be grayed out?

Some people suggested selected area to be non grayed out.
I think that makes sense, such that you see what you are grabbing.

**Comment by alimpfard** [comment]
Yeah, I expect that the selected area shouldn't be grayed out, but it looks like this image does the inverse?
>

**Comment by abyesilyurt** [comment]
> Yeah, I expect that the selected area shouldn't be grayed out, but it looks like this image does the inverse?
> 
> > 
> 
> Update: Apparently my eyes are tricking me here

I picked Blue over Gray for the background, otherwise the contrast was very low.

**Comment by alimpfard** [comment]
:thinking:
maybe try white or the widget base colour (`palette().base()`)?

File: Userland/Utilities/shot.cpp [hunk 13]
```diff
@@ -1,32 +1,103 @@
...
 #include <LibGUI/WindowServerConnection.h>
 #include <LibGfx/PNGWriter.h>
 #include <unistd.h>
 
+class SelectableLayover final : public GUI::Widget {
+    C_OBJECT(SelectableLayover)
+public:
+    SelectableLayover(GUI::Window* window)
+        : m_window(window)
+    {
```
Comment (gunnarbeutner) [inline Userland/Utilities/shot.cpp:27]:
```suggestion
    {
        set_override_cursor(Gfx::StandardCursor::Crosshair);
```

Comment (gunnarbeutner) [reply Userland/Utilities/shot.cpp:27]: That way you get a crosshair cursor for the selection window:

image

Comment (abyesilyurt) [reply Userland/Utilities/shot.cpp:29]: Cool!

**Comment by abyesilyurt** [comment]
> maybe try white or the widget base colour (`palette().base()`)?

palette().threed_highlight() gives the best contrast overall considering the other themes.

**Comment by abyesilyurt** [comment]
This is how it looks now.

image

**Comment by alimpfard** [comment]
:thinking: that does look nice, maybe if you bump the opacity a little bit, it'd make that a little more visible?

**Comment by alimpfard** [comment]
Also, I just noticed that the screenshot window doesn't have a title :P 
Please give it a title lol

**Comment by abyesilyurt** [comment]
>  that does look nice, maybe if you bump the opacity a little bit, it'd make that a little more visible?

That is the problem. Opacity is global over window, if I increase the opacity, the transparent part is effected at the same time.

**Comment by abyesilyurt** [comment]
> Also, I just noticed that the screenshot window doesn't have a title :P
> Please give it a title lol

:D "shot"?

**Comment by alimpfard** [comment]
> That is the problem. Opacity is global over window, if I increase the opacity, the transparent part is effected at the same time.

Then let's not block this PR, that's a separate issue anyway.

> "shot"?

Sure :)

**Comment by abyesilyurt** [comment]
Tests are passing, can we merge?

**Comment by abyesilyurt** [comment]
> thinking that does look nice, maybe if you bump the opacity a little bit, it'd make that a little more visible?

I opened an issue for this, #8577.

**Comment by alimpfard** [comment]
Let's go with it
